packages/evm/README.md

```diff
@@ -422,28 +422,58 @@ Additional log selections can be added with a comma separated list (no spaces).
 
 ### Internal Structure
 
-The EVM processes state changes at many levels.
-
-- **runCall**
-  - checkpoint state
-  - transfer value
-  - load code
-  - runCode
-  - materialize created contracts
-  - revert or commit checkpoint
-- **runCode**
-  - iterate over code
-  - run op codes
-  - track gas usage
-- **OpFns**
-  - run individual op code
-  - modify stack
-  - modify memory
-  - calculate fee
-
-The opFns for `CREATE`, `CALL`, and `CALLCODE` call back up to `runCall`.
-
-TODO: this section likely needs an update.
+The EVM processes state changes through a hierarchical flow of execution:
+
+#### Top Level: Message Execution (`runCall`)
+The `runCall` method handles the execution of messages, which can be either contract calls or contract creations:
+- Creates a checkpoint in the state
+- Sets up the execution environment (block context, transaction origin, etc.)
+- Manages account nonce updates
+- Handles value transfers between accounts
+- Delegates to either `_executeCall` or `_executeCreate` based on whether the message has a `to` address
+- Processes any errors or exceptions
+- Manages selfdestruct sets and created contract addresses
+- Commits or reverts state changes based on execution result
+- Triggers events (`beforeMessage`, `afterMessage`)
+
+#### Code Execution (`runCode` / `runInterpreter`)
+The `runCode` method is responsible for bytecode execution:
+- Sets up a message context for code execution
+- The `runInterpreter` method then:
+  - Creates an execution environment (address, caller, value, etc.)
+  - Instantiates an Interpreter with the message context
+  - Handles EOF code execution if applicable (EIP-3540)
+  - Processes execution results
+
+#### Bytecode Processing (Interpreter)
+The Interpreter class is the core bytecode processor:
+- Manages execution state (program counter, stack, memory, gas)
+- Executes a loop that:
+  - Analyzes jump destinations
+  - Fetches the next opcode
+  - Calculates gas costs (static and dynamic)
+  - Executes the opcode handler
+  - Updates the program counter
+  - Emits step events for debugging/tracing
+- Handles stack, memory, and storage operations
+- Processes call and creation operations by delegating back to the EVM
+
+#### Opcode Functions
+Each opcode has an associated handler function that:
+- Validates inputs
+- Calculates dynamic gas costs
+- Performs the opcode's logic (stack operations, memory operations, etc.)
+- Updates the EVM state
```

Comment:
```suggestion
- Updates the EVM state
- The program counter is incremented in between the execution of the gas handler and opcode logic handler functions, this should be considered e.g. if parsing immediate input parameters
```

Reply:
This was a tidbit that was discovered while working on opcode implementations for [eip 6690](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6690.md)